lib/ansible/parsing/vault/__init__.py

```diff
@@ -500,58 +503,42 @@ def rekey_file(self, filename, new_password):
         os.chown(filename, prev.st_uid, prev.st_gid)
 
     def read_data(self, filename):
+        data = None
 
-        try:
-            if filename == '-':
+        if filename == '-':
+            try:
                 data = sys.stdin.read()
-            else:
-                with open(filename, "rb") as fh:
-                    data = fh.read()
+            except Exception as e:
+                raise AnsibleError(str(e))
+            return
+
+        try:
+            with open(filename, "rb") as fh:
+                data = fh.read()
         except Exception as e:
             raise AnsibleError(str(e))
 
         return data
 
     # TODO: add docstrings for arg types since this code is picky about that
-    def write_data(self, data, filename, shred=True):
+    def write_data(self, data, filename):
         """write data to given path
 
         :arg data: the encrypted and hexlified data as a utf-8 byte string
         :arg filename: filename to save 'data' to.
-        :arg shred: if shred==True, make sure that the original data is first shredded so
-        that is cannot be recovered.
         """
         # FIXME: do we need this now? data_bytes should always be a utf-8 byte string
         b_file_data = to_bytes(data, errors='strict')
 
         if filename == '-':
             sys.stdout.write(b_file_data)
-        else:
-            if os.path.isfile(filename):
-                if shred:
-                    self._shred_file(filename)
-                else:
-                    os.remove(filename)
-            with open(filename, "wb") as fh:
-                fh.write(b_file_data)
-
-    def shuffle_files(self, src, dest):
-        prev = None
-        # overwrite dest with src
-        if os.path.isfile(dest):
-            prev = os.stat(dest)
-            # old file 'dest' was encrypted, no need to _shred_file
-            os.remove(dest)
-        shutil.move(src, dest)
-
-        # reset permissions if needed
-        if prev is not None:
-            # TODO: selinux, ACLs, xattr?
-            os.chmod(dest, prev.st_mode)
-            os.chown(dest, prev.st_uid, prev.st_gid)
+            return
+
+        with open(filename, "wb") as fh:
+            fh.write(b_file_data)
```

Comment:
Was trying to avoid the gap between the tmp file rename and resetting perms/attributes in shuffle_files. This does allow a read of a partially written file, but it also keeps file perms in place.

I like the idea of getting perms sorted on the tmpfile before the rename  (as mentioned at https://github.com/ansible/ansible/issues/18537). Though still a small perm race there (dest file perms could change in the gap between the setting/copying file perms from dest file to the temp file and when the rename happens) .